commonmark/src/main/java/org/commonmark/internal/InlineParserImpl.java

```diff
@@ -163,44 +143,48 @@ public void parse(String content, Node block) {
         mergeChildTextNodes(block);
     }
 
+    void setPosition(Position position) {
+        index = position.getIndex();
+    }
+
     void reset(String content) {
         this.input = content;
         this.index = 0;
         this.lastDelimiter = null;
         this.lastBracket = null;
     }
 
-
-    private Text text(String text, int beginIndex, int endIndex) {
-        return new Text(text.substring(beginIndex, endIndex));
-    }
-
     private Text text(String text) {
         return new Text(text);
     }
 
     /**
-     * Parse the next inline element in subject, advancing input index.
+     * Parse the next inline element in subject, advancing our position.
      * On success, return the new inline node.
      * On failure, return null.
      */
     private Node parseInline(Node previous) {
-        char c = peek();
+        Scanner scanner = scanner();
+        char c = scanner.peek();
         if (c == '\0') {
             return null;
         }
 
+        List<InlineContentParser> inlineParsers = this.inlineParsers.get(c);
+        if (inlineParsers != null) {
+            for (InlineContentParser inlineParser : inlineParsers) {
```

Comment:
hmm I applied a similar idea here https://github.com/atlassian/commonmark-java/pull/178/files#diff-abb5406ad2ea9b67e29591c9ffb7d35bR45

One challenge that I saw was in case we do have more than one `InlineContentParser` using the same start symbol. We don't have how to know in the current position which one will find the "end character" to decide which one is the valid one. That's why I kept one "InlineContentParser" active list and one array where I could notice which node could be fully parsed.

I can see that we have one example which has the same start symbol
`this.inlineParsers.put('<', Arrays.asList(new AutolinkInlineParser(), new HtmlInlineParser()));` 
For `AutolinkInlineParser` we need from the start symbol until the close `>` but we may not have, right? Thinking when we add here the extension of how to handle this in case the client code also adds a third InliineParser for `<`.. Any idea?

Reply:
Not sure what you mean, can you explain with an example? With this, the first inline parser that returns a result will win. To decide, the inline parser can look at as much of the input as it wants (until the end of the block).

For custom inline parsers, they should probably come first, so they can modify the core syntax. In this case, if a custom inline parser for `<` was registered, it could do whatever it wants, before the core inline parsers would have a chance.

Reply:
Summaryzing what I tried to say is that the number of interactions will be multiplied by the number of (Node)Parsers.

Let me try to put the example below:

```
Taking the defaults parsers:
'<'
  AutolinkInlineParser -> scanner.find('>') > 0
    URI.matcher(text).matches()
    or
    EMAIL.matcher(text).matches()
   
  HtmlInlineParser
    Regex
    or
    to look for some "end character" in the line
    


Parsing the line below:
  Some dummy text < anybloh text here>
                  ^
                  -------------------^
                    AutolinkInlineParser,
                      return None
                  ^
                  -------------------^
                    HtmlInlineParser
                      return None    
``` 

if we add extensions for `<` we would have `(N * extensions start by '<' )` interacting in this line. But thinking better we may not have how to avoid it.


The model that I tried in that other PR it kind of a little bit different for that one that you took. Taking the same example the algorithm is:

```
Parsing the line below:
  Some dummy text < anybloh text here>
                  ^
                  AutolinkInlineParser/HtmlInlineParser active
                  -------------------^
                                     AutolinkInlineParser.isComplete()?
                                      return None;
                                     HtmlInlineParser.isComplete()?
                                      return None
```

Frankly speaking, I don't know which one is the best. It seems the approach that you took we can avoid holding multiple parsers watching the line at the same time. So, I'm with you. :)

Reply:
Ah yeah, I know what you mean now. I think given that it works the same way as before this change, it should be ok. I like the model with this change, it's simple to understand and works the same way as block parsers.